osu! Difficulty Aim Overhaul: base change

Xexxar opened the pull request:
subset of https://github.com/ppy/osu/pull/14767 featuring focuses 1, 2, 3, 4, 6.

Xexxar committed:
base change of aim refactor, isolated

bdach committed:
Update SR values in tests

bdach commented on osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:
```diff
@@ -37,6 +37,11 @@ public class OsuDifficultyHitObject : DifficultyHitObject
         public double? Angle { get; private set; }
 
+        /// <summary>
+        /// Milliseconds elapsed since the start time of the previous <see cref="OsuDifficultyHitObject"/>, with a minimum of 50ms.
```
Wait, isn't it
```suggestion
        /// Milliseconds elapsed since the start time of the previous <see cref="OsuDifficultyHitObject"/>, with a minimum of 25ms.
```
as per this lower down?
https://github.com/ppy/osu/blob/671fe772136671189cd404d6d9f723dc698050ad/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs#L51-L52

bdach commented:
Aside from the above the parts that I can check look OK.

FamousPig commented:
The original pr adjusted the aim scaling with accuracy in ` OsuPerformanceCalculator.cs` which is omitted here, changing the resulting values. Is this intentional?

Xexxar commented:
> The original pr adjusted the aim scaling with accuracy in ` OsuPerformanceCalculator.cs` which is omitted here, changing the resulting values. Is this intentional?

This is an oversite, will be fixed.

Xexxar committed:
readded missing aim nerf for low acc

Xexxar committed:
Merge branch 'aim-refactor-base' of https://github.com/emu1337/osu into aim-refactor-base

Xexxar committed:
Update osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

stanriders commented on osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:
```diff
@@ -14,51 +16,97 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+
+        private double calcWideAngleBonus(double angle)
+        {
+            if (angle < Math.PI / 3)
```
Maybe add comments or even conversion to degrees? It'd be a bit more convenient to read

stanriders commented on osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:
```diff
@@ -14,51 +16,97 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
 
-            double jumpDistanceExp = applyDiminishingExp(osuCurrent.JumpDistance);
-            double travelDistanceExp = applyDiminishingExp(osuCurrent.TravelDistance);
+                if (osuCurrObj.StrainTime + 10 < osuPrevObj.StrainTime && osuPrevObj.StrainTime > osuLastObj.StrainTime + 10)
```
I think it'd be better to `const` this number (10) as a `rhythm_change_threshold` or something since it's being used in multiple places

stanriders commented on osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:
```diff
@@ -14,51 +16,97 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+                    double wideAngleBonus = calcWideAngleBonus(angle);
+                    double acuteAngleBonus = calcAcuteAngleBonus(angle);
+
+                    if (osuCurrObj.StrainTime > 100)
```
This should probably state somewhere how fast 100 straintime is

Xexxar replied:
done.

stanriders commented on osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:
```diff
@@ -14,51 +16,97 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
-                result + (jumpDistanceExp + travelDistanceExp + Math.Sqrt(travelDistanceExp * jumpDistanceExp)) / Math.Max(osuCurrent.StrainTime, timing_threshold),
-                (Math.Sqrt(travelDistanceExp * jumpDistanceExp) + jumpDistanceExp + travelDistanceExp) / osuCurrent.StrainTime
-            );
+                aimStrain += rhythmBonus * rhythm_variance_multiplier; // add in rhythm velocity.
```
Honestly I'm not sure if aim should reward rhythm at all, this should be covered by rhythm complexity alone. Aiming rhythm changes isn't hard by itself, reading might be, but it's out of scope of aim skill

emu1337 commented on osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:
```diff
@@ -14,51 +16,97 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+                return Math.Pow(Math.Sin(1.5 * (2 * Math.PI / 3 - angle)), 2);
+
+            return 0;
+        }
```
The wide angle bonus peaks at 120 degrees and the acute one peaks at 60, which is fine, however if they gain too much compared to the entire range of wide/acute angles it could present a problem because these particular angles buff specific patterns too much. I would try flatter curves here that also reache the peak faster, for example at 75/105 degrees.

As it currently stands the bonus difference between a 100 deg wide angle jump and a 120 deg one is about 25% which seems unrealistic to me.

emu1337 replied:
A kind user by the name of **joz** provided some [data](https://cdn.discordapp.com/attachments/886986744090734682/893153129707733082/unknown.png) on the difficulty of wide angles. This was gathered through analyzing actual replays with machine learning, and as far as I'm concerned it makes more sense than the implementation here.

The findings are that wide angles increase difficulty by up to 20%-25% regardless of bpm, and the rate of increase resembles a sigmoid function.

Xexxar replied:
If we have an implementation of this somewhere, this would likely be a great replacement to the existing wide angle calculation.

Xexxar committed:
merged ppy/master

Xexxar committed:
Merge remote-tracking branch 'ppy/master' into aim-refactor-base

Xexxar committed:
updated to newly refactored aim

Xexxar committed:
removed the merge conflict markers

Xexxar committed:
removed residual code leftover

Xexxar committed:
adjusted constant to 1.5 from 2.0 for acute

smoogipoo committed:
Merge branch 'master' into aim-refactor-base

smoogipoo commented:
A few NaN's / Infinity values being generated here:
```
/b/184846 -> Infinity
/b/246433 -> NaN
/b/572525 -> Infinity
/b/646391 -> Infinity
/b/1034190 -> Infinity
/b/1073867 -> Infinity
/b/1089618 -> Infinity
/b/1089619 -> Infinity
/b/1133255 -> NaN
/b/1151178 -> NaN

... (And many more)
```
Goes without saying but even if these are loved maps and thus probably exhibit some edge cases, NaN/Infinity values are not allowed.

Xexxar committed:
resolved nans and added stacked slider fix

smoogipoo commented:
[Spreadsheet](https://docs.google.com/spreadsheets/d/1bAFleOYszZ4OjNYvb5u0VhOAQ_091mgLORIEAwTdwIQ/edit#gid=0) of top SR/PP changes.

Will look at merging this with some approvals from the pp committee.

Xexxar committed:
Merge remote-tracking branch 'ppy/master' into aim-refactor-base

Xexxar committed:
removed ppCalc changes and sliderabuseChecks

Xexxar committed:
identified case with spinner where / 0 could theoretically be possible

Xexxar committed:
reverted overwritten file

Xexxar committed:
updated to add nerf for repeated acute angles

Xexxar committed:
add in stanrs review

Xexxar committed:
merge in ppy/master

smoogipoo committed:
Resolve tests

smoogipoo committed:
Apply styling changes

smoogipoo committed:
Refactor xmldocs

smoogipoo committed:
Rename variables for readability

In particular, "last" vs "prev" can get confusing because they mean the
same thing.

smoogipoo committed:
Refactor documentation around velocity calculations

smoogipoo committed:
Refactor variables for readability in angle calculations

Xexxar commented on osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:
```diff
@@ -71,11 +87,15 @@ private void setDistances()
             Vector2 lastCursorPosition = getEndCursorPosition(lastObject);
 
             JumpDistance = (BaseObject.StackedPosition * scalingFactor - lastCursorPosition * scalingFactor).Length;
+            MovementDistance = Math.Min(JumpDistance, MovementDistance);
```
These could be set to be equal to straintime for circles. In fact that would probably more accurately match the definition.

smoogipoo replied:
Yeah that's the thing, it's going against the xmldocs which was unexpected to me at first read.

So basically:
```
MovementDistance = JumpDistance;
MovementTime = StrainTime;
```
For hitcircles?

Xexxar replied:
I have no problem with that definition.

Xexxar replied:
For all intents and purposes, movement distance could seen as the same thing as junp distance. The idea here in the definition is that there is a leniency of the slider where your movement can be either from the minimum end point of the slider travel distance, or from the actual tail end. This way we can better assess the true observed velocity for the movement to the next note. Perhaps a better way to put it is that when you take into consideration the movement from the slider to the next note, lazy end is not always where its actually easiest to end if the movement after the slider is in the direction of the slider.

Xexxar replied:
The slider augmentation here is that we treat sliders as two seperate velocities, rather than 1. The core benefit of viewing things this way is that we buff sliders by assessing two movements instead of one. 

The reason why I added this leniency is because without it, slider jumps of a certain design become overweighted as the secondary velocity to the following note is seen as faster than the actual difficulty in gameplay is. This is due to the inaccurate real cursor position by using lazy position exclusively.

smoogipoo replied:
Yeah, sure. Again just understand that this will nerf these sort of patterns on the other side of the spectrum:
![image](https://user-images.githubusercontent.com/1329837/139890541-2fcb2af3-1f90-4456-b02d-887bea160ada.png)
(A player will not follow either of those)

smoogipoo commented:
I've made a few refactorings for my non-mathy brain to handle it reading through things.

Xexxar commented on osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:
```diff
@@ -14,53 +15,92 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
         private double strainValueOf(DifficultyHitObject current)
         {
-            if (current.BaseObject is Spinner)
+            if (current.BaseObject is Spinner || Previous.Count <= 1 || Previous[0].BaseObject is Spinner)
```
The check for previous is needed due to how movement time and traveltime are zero when the next object is a spinner. It doesnt matter if previous[1] is a spinner since we dont look at that objs times or distances.

smoogipoo committed:
Merge branch 'master' into aim-refactor-base

smoogipoo committed:
Document special MovementDistance for sliders, ensure always has a value
